runtime/src/main/cpp/Memory.cpp

```diff
@@ -2252,6 +2253,96 @@ void FreezeSubgraph(ObjHeader* root) {
 #endif
 }
 
+void deepFrozenCopyTo(ObjHeader* to, ObjHeader* from, ObjHeaderMap* translationMapm, ObjHeader** firstBlocker);
+
+bool cannotClone(ObjHeader* original) {
+  if (original == nullptr)
+    return false;
+  if ((original->type_info()->flags_ & TF_NON_CLONEABLE) != 0)
+    return true;
+  if (original->has_meta_object() && ((original->meta_object()->flags_ & MF_NEVER_FROZEN) != 0))
+    return true;
+  return false;
```

Comment:
It's also not correct to clone Kotlin object with non-trivial Objective-C counterpart (e.g. for instance of Objective-C subclass of Kotlin class).